File: frontend/src/api/assets.js
import { get, post, put, del } from "./http.js";

export function listAssets({
  scope = "all",
  projectId = "",
  assetType = "",
  category,
  enabledOnly = false,
  limit = 200,
  offset = 0,
} = {}) {
  const params = new URLSearchParams();
  params.set("scope", scope);
  if (projectId) params.set("project_id", projectId);
  if (assetType) params.set("asset_type", assetType);
  if (category !== undefined && category !== null) params.set("category", category);
  if (enabledOnly) params.set("enabled_only", "true");
  params.set("limit", String(limit));
  params.set("offset", String(offset));
  return get(`/api/assets?${params.toString()}`);
}

export function searchAssets(payload) {
  return post(`/api/assets/search`, payload);
}

export function getAsset(assetId, { projectId } = {}) {
  const qs = projectId ? `?project_id=${encodeURIComponent(projectId)}` : "";
  return get(`/api/assets/${assetId}${qs}`);
}

export function createAsset(payload) {
  return post(`/api/assets`, payload);
}

export function updateAsset(assetId, payload) {
  return put(`/api/assets/${assetId}`, payload);
}

export function deleteAsset(assetId, { scope = "global", projectId = "" } = {}) {
  const params = new URLSearchParams({ scope });
  if (projectId) params.set("project_id", projectId);
  return del(`/api/assets/${assetId}?${params.toString()}`);
}

export function batchToggleAssets(payload) {
  return post(`/api/assets/batch-toggle`, payload);
}

export function batchCategorizeAssets(payload) {
  return post(`/api/assets/batch-categorize`, payload);
}

export function startStyleExtraction(payload) {
  return post(`/api/assets/style-extract`, payload);
}

export function getStyleExtractionStatus(taskId) {
  return get(`/api/assets/style-extract/${taskId}`);
}

// ---- 入库 Agent ----
export function startAssetIngestion(payload) {
  return post(`/api/assets/ingest`, payload);
}

export function getAssetIngestionStatus(taskId) {
  return get(`/api/assets/ingest/${taskId}`);
}

// ---- 统一聚合视图（跨 silo） ----
export function listUnifiedAssets({
  projectId = "",
  sources = [],
  entityTypes = [],
  scope = "",
  q = "",
  page = 1,
  pageSize = 50,
} = {}) {
  const params = new URLSearchParams();
  if (projectId) params.set("project_id", projectId);
  if (sources.length) params.set("source", sources.join(","));
  if (entityTypes.length) params.set("entity_type", entityTypes.join(","));
  if (scope) params.set("scope", scope);
  if (q) params.set("q", q);
  params.set("page", String(page));
  params.set("page_size", String(pageSize));
  return get(`/api/unified-assets?${params.toString()}`);
}

export function getUnifiedFacets(projectId = "") {
  const qs = projectId ? `?project_id=${encodeURIComponent(projectId)}` : "";
  return get(`/api/unified-assets/facets${qs}`);
}

export function searchGlobalAssets({ q, projectId = "", sources = [], entityTypes = [], limit = 50 } = {}) {
  const params = new URLSearchParams({ q });
  if (projectId) params.set("project_id", projectId);
  if (sources.length) params.set("source", sources.join(","));
  if (entityTypes.length) params.set("entity_type", entityTypes.join(","));
  params.set("limit", String(limit));
  return get(`/api/unified-assets/search?${params.toString()}`);
}

export function reindexUnifiedAssets(projectId) {
  return post(`/api/unified-assets/reindex`, { project_id: projectId });
}

File: backend/app/api/assets.py
"""资产库 REST API。"""

from __future__ import annotations

import traceback

from flask import jsonify, request

from . import assets_bp
from ..models.task import TaskManager
from ..services.assets.assets_service import AssetsService
from ..services.assets.assets_storage import GLOBAL_SCOPE, PROJECT_SCOPE
from ..services.assets.style_extractor import StyleExtractor


def _service() -> AssetsService:
    return AssetsService()


def _err(exc: Exception, status: int = 500):
    return jsonify({
        "success": False,
        "error": str(exc),
        "traceback": traceback.format_exc(),
    }), status


def _validate_scope(scope: str) -> str:
    if scope not in (GLOBAL_SCOPE, PROJECT_SCOPE):
        raise ValueError(f"scope 必须是 global 或 project，收到 {scope!r}")
    return scope


# ----------------------------------------------------------------------
# Listing & search
# ----------------------------------------------------------------------


@assets_bp.route("", methods=["GET"])
def list_assets():
    try:
        scope = request.args.get("scope", "all")
        project_id = request.args.get("project_id") or None
        asset_type = request.args.get("asset_type") or None
        category = request.args.get("category")
        enabled_only = request.args.get("enabled_only", "false").lower() == "true"
        limit = int(request.args.get("limit", "200"))
        offset = int(request.args.get("offset", "0"))

        svc = _service()
        if scope == "all":
            data = svc.list_merged(
                project_id=project_id,
                asset_type=asset_type,
                category=category,
                enabled_only=enabled_only,
                limit=limit,
            )
        else:
            _validate_scope(scope)
            data = svc.list(
                scope=scope, project_id=project_id,
                asset_type=asset_type, category=category,
                enabled_only=enabled_only, limit=limit, offset=offset,
            )
        return jsonify({"success": True, "data": data})
    except Exception as exc:
        return _err(exc)


@assets_bp.route("/search", methods=["POST"])
def search_assets():
    try:
        body = request.get_json() or {}
        query = body.get("query", "")
        if not query.strip():
            return jsonify({"success": False, "error": "query 不能为空"}), 400
        scope = body.get("scope", "all")
        project_id = body.get("project_id") or None
        asset_type = body.get("asset_type") or None
        category = body.get("category")
        enabled_only = bool(body.get("enabled_only", True))
        limit = int(body.get("limit", 20))

        svc = _service()
        if scope == "all":
            data = svc.search_merged(
                query, project_id=project_id, asset_type=asset_type,
                category=category, enabled_only=enabled_only, limit=limit,
            )
        else:
            _validate_scope(scope)
            data = svc.search(
                query, scope=scope, project_id=project_id,
                asset_type=asset_type, category=category,
                enabled_only=enabled_only, limit=limit,
            )
        return jsonify({"success": True, "data": data})
    except Exception as exc:
        return _err(exc)


# ----------------------------------------------------------------------
# CRUD
# ----------------------------------------------------------------------


@assets_bp.route("/<asset_id>", methods=["GET"])
def get_asset(asset_id: str):
    try:
        project_id = request.args.get("project_id") or None
        asset = _service().find(asset_id, project_id=project_id)
        if not asset:
            return jsonify({"success": False, "error": "未找到资产"}), 404
        return jsonify({"success": True, "data": asset})
    except Exception as exc:
        return _err(exc)


@assets_bp.route("", methods=["POST"])
def create_asset():
    try:
        body = request.get_json() or {}
        scope = _validate_scope(body.get("scope", GLOBAL_SCOPE))
        asset = _service().create(
            scope=scope,
            asset_type=body["asset_type"],
            title=body["title"],
            project_id=body.get("project_id"),
            category=body.get("category", ""),
            summary=body.get("summary", ""),
            content=body.get("content", ""),
            payload=body.get("payload"),
            tags=body.get("tags"),
            source_kind=body.get("source_kind", "manual"),
            source_ref=body.get("source_ref", ""),
            enabled=bool(body.get("enabled", True)),
            pinned=bool(body.get("pinned", False)),
        )
        return jsonify({"success": True, "data": asset})
    except Exception as exc:
        return _err(exc)


@assets_bp.route("/<asset_id>", methods=["PUT"])
def update_asset(asset_id: str):
    try:
        body = request.get_json() or {}
        scope = _validate_scope(body.pop("scope", GLOBAL_SCOPE))
        project_id = body.pop("project_id", None)
        asset = _service().update(asset_id, scope=scope, project_id=project_id, **body)
        return jsonify({"success": True, "data": asset})
    except Exception as exc:
        return _err(exc)


@assets_bp.route("/<asset_id>", methods=["DELETE"])
def delete_asset(asset_id: str):
    try:
        scope = _validate_scope(request.args.get("scope", GLOBAL_SCOPE))
        project_id = request.args.get("project_id") or None
        ok = _service().delete(asset_id, scope=scope, project_id=project_id)
        return jsonify({"success": ok})
    except Exception as exc:
        return _err(exc)


# ----------------------------------------------------------------------
# Batch
# ----------------------------------------------------------------------


@assets_bp.route("/batch-toggle", methods=["POST"])
def batch_toggle():
    try:
        body = request.get_json() or {}
        scope = _validate_scope(body.get("scope", GLOBAL_SCOPE))
        n = _service().batch_set_enabled(
            body.get("asset_ids") or [],
            bool(body.get("enabled", True)),
            scope=scope, project_id=body.get("project_id"),
        )
        return jsonify({"success": True, "data": {"updated": n}})
    except Exception as exc:
        return _err(exc)


@assets_bp.route("/batch-categorize", methods=["POST"])
def batch_categorize():
    try:
        body = request.get_json() or {}
        scope = _validate_scope(body.get("scope", GLOBAL_SCOPE))
        n = _service().batch_set_category(
            body.get("asset_ids") or [],
            body.get("category", ""),
            scope=scope, project_id=body.get("project_id"),
        )
        return jsonify({"success": True, "data": {"updated": n}})
    except Exception as exc:
        return _err(exc)


# ----------------------------------------------------------------------
# Style extraction
# ----------------------------------------------------------------------


@assets_bp.route("/style-extract", methods=["POST"])
def style_extract():
    try:
        body = request.get_json() or {}
        text = body.get("text") or ""
        title = (body.get("title") or "").strip()
        if not text.strip() or not title:
            return jsonify({"success": False, "error": "text 与 title 必填"}), 400
        task_id = StyleExtractor().extract_background(
            text,
            title=title,
            category=body.get("category", ""),
            tags=body.get("tags") or [],
            target_chunk_chars=int(body.get("target_chunk_chars") or 3000),
            max_chunks=int(body.get("max_chunks") or 30),
        )
        return jsonify({"success": True, "data": {"task_id": task_id}})
    except Exception as exc:
        return _err(exc)


@assets_bp.route("/style-extract/<task_id>", methods=["GET"])
def style_extract_status(task_id: str):
    try:
        task = TaskManager().get_task(task_id)
        if not task:
            return jsonify({"success": False, "error": "任务不存在"}), 404
        return jsonify({"success": True, "data": task.to_dict()})
    except Exception as exc:
        return _err(exc)
